### ISSUE TITLE
Numerical issues with NLshrink

### ISSUE THREAD

**Issue by tlienart** [mateuszbaran/CovarianceEstimation.jl#38]
**Short summary**: in the case `n<p` I think their method leads to numerical problems.

I was in the process of constructing benchmarks to evaluate the speedup gains we get over other libraries (5x over sklearn, up to 30x over corpcor). In the process of experimenting with the matlab version, I coded a translation of their code into julia and then while comparing I ran into troubles.

The main thing is that their matlab code does not center the matrix. If you do center the matrix and the matrix has `n<p`, you lose one in rank, which means that one of the lambdas in `λ = λ[max(1, p-n+1):p]` is going to be (near)zero. 

Easier to show than write:

```julia
using CovarianceEstimation
using Random:seed!
Random.seed!(1234)
X = randn(20, 10); # tall case
Xc = CovarianceEstimation.centercols(X);
lambda, U = eigen(Xc'*Xc)
```

That's the nice case, none of the lambdas are zero

```
eigenvalues:
10-element Array{Float64,1}:
  1.0448369601469278
  6.873678897981916
  8.649561805189752
 13.603590792009618
 16.598666758476913
 23.97710606404871
 25.616137487782645
 29.03009163529542
 30.90358831104026
 46.12381521585943
```

Now the not-nice case:

```julia
using CovarianceEstimation
using Random:seed!
Random.seed!(1234)
X = randn(10, 20); # fat case
Xc = CovarianceEstimation.centercols(X);
lambda, U = eigen(Xc'*Xc)
```

which gives

```
eigenvalues:
20-element Array{Float64,1}:
 -3.645626550750355e-15
 -3.577709959236032e-15
 -1.831741337515713e-15
 -1.3950392087987232e-15
 -1.2686174094960056e-15
 -1.2147395397384545e-15
 -6.18254383688614e-17
  6.838205226971953e-16
  9.702880122503415e-16
  1.380856443067124e-15
  3.649556353209685e-15
  4.3412780730655305         # 1
  8.827158115761128
 11.0559161018662 
 16.32761764211215
 19.533587171231783
 20.82149090505307
 29.69787977758131
 33.75273902860496
 51.490074671031245 # 9
```

So here there's only 9 eigenvalues that are non zero. This matters a lot because following their code and taking

```julia
n, p = size(X)
lambda = lambda[max(1, p-n+1):p]
```

one of the eigenvalues is zero

```
3.649556353209685e-15
  4.3412780730655305
  8.827158115761128
 11.0559161018662
 16.32761764211215
 19.533587171231783
 20.82149090505307
 29.69787977758131
 33.75273902860496
 51.490074671031245
```

This propagates through `L` then `H` etc with infinities and NaNs being thrown around. 

### what we should do

1. see whether there's an easy work around for this case where, for instance, we can make sense of one of the lambda being zero and propagate it through "safely"
2. we should probably contact the authors 

If you agree with the analysis, ping me over Slack and we can write the email there.

**Comment by mateuszbaran** [comment]
Oh right, this is the case I didn't actually finish testing. I've taken another look at their paper and it looks like they didn't really taken this into account. In general, even nonzero but close-to-zero eigenvalues significantly impact the result (`Hftilde0` depends on the mean of inverses of eigenvalues). But it looks like it should be relatively safe to just replace there `n` by `n-1` (or by rank of the covariance matrix). It would probably be a good idea to contact the authors though.

**Comment by tlienart** [comment]
FWIW, here's a very simple example ran in Octave and showing the issue. 

```matlab
function sigmatilde=analytical_shrinkage(X)
    % extract sample eigenvalues sorted in ascending order and eigenvectors
    [n, p] = size(X);
    meanx=mean(X);
    X=X-meanx(ones(n,1),:);
    % important: sample size n must be >= 12
    sample = (X' * X) ./ n;
    [u, lambda] = eig(sample, 'vector');
    [lambda, isort] = sort(lambda);
    u = u(:, isort);
    % compute analytical nonlinear shrinkage kernel formula
    lambda = lambda(max(1,p-n+1):p);
    L = repmat(lambda,[1 min(p,n)]);
    h = n^(-1/3);
    H = h*L';
    x = (L-L')./H;
    ftilde = (3/4/sqrt(5))*mean(max(1-x.^2./5,0)./H,2);
    Hftemp = (-3/10/pi)*x+(3/4/sqrt(5)/pi)*(1-x.^2./5).*log(abs((sqrt(5)-x)./(sqrt(5)+x)));
    Hftemp(abs(x)==sqrt(5)) = (-3/10/pi)*x(abs(x)==sqrt(5));
    Hftilde = mean(Hftemp./H, 2);
    if p<=n
      dtilde = lambda./((pi*(p/n)*lambda.*ftilde).^2+(1-(p/n)-pi*(p/n)*lambda.*Hftilde).^2);
    else
      Hftilde0 = (1/pi)*(3/10/h^2+3/4/sqrt(5)/h*(1-1/5/h^2)*log((1+sqrt(5)*h)/(1-sqrt(5)*h)))*mean(1./lambda);
      dtilde0 = 1/(pi*(p-n)/n*Hftilde0);
      dtilde1 = lambda./(pi^2*lambda.^2.*(ftilde.^2+Hftilde.^2));
      dtilde = [dtilde0*ones(p-n,1);dtilde1];
    end
    sigmatilde=u*diag(dtilde)*u';
end

rand("seed", 123)
X = randn(3, 5);
mu = mean(X);
Xc = X - mu(ones(3, 1), :);
St = analytical_shrinkage(Xc);
norm(St) % 1e-17 in Octave 
```

**Comment by mateuszbaran** [comment]
Hmm... at the beginning of section 2.1 they assume that samples mean-zero iid. But when we don't know the mean we need to estimate it as well, and we'd have to subtract it anyway.

**Comment by tlienart** [comment]
ok it makes sense that if these conditions are met then their algorithm is fine. I wonder whether the "generalisation" where you'd consider at most max(1, p-n+1) *nonzero* eigenvalues is sound or not.

**Comment by tlienart** [comment]
fix suggested in #46
